### ISSUE TITLE
Generate structure of objects instead of hash when parsing

### ISSUE THREAD

**Issue by herwinw** [pganalyze/pg_query#96]
Be warned: this is a feature request that is going to need a heavy restructuring of the code.

Currently, parsing a query results in a `PgQuery` object that has a few methods and mainly relies on a heavy hash/array-structure.

```ruby
irb(main):002:0> PgQuery.parse('select 1').public_methods(false).sort
=> [:aliases, :cte_names, :ddl_tables, :deep_dup, :deparse, :dml_tables, :filter_columns, :fingerprint, :param_refs, :parsetree, :query, :select_tables, :tables, :tables_with_types, :tree, :truncate, :warnings]
irb(main):003:0> PgQuery.parse('select 1')
=> #<PgQuery:0x572748ac @query="select 1", @tree=[{"RawStmt"=>{"stmt"=>{"SelectStmt"=>{"targetList"=>[{"ResTarget"=>{"val"=>{"A_Const"=>{"val"=>{"Integer"=>{"ival"=>1}}, "location"=>7}}, "location"=>7}}], "op"=>0}}}}], @warnings=[], @tables=nil, @aliases=nil, @cte_names=nil>
```

Most operations on the data need to be performed via the tree structure. An example from the README:

```ruby
# Modify the parse tree in some way
parsed_query.tree[0]['SelectStmt']['fromClause'][0]['RangeVar']['relname'] = 'other_users'
```

I was reworking some unit tests earlier today that tried to validate the output of a query generator. The specific query tried to check the ORDER BY clause, which was done by combining `assert_includes` with the generated query and a hardcoded order by string. Since we don't really care about things like the layout or the casing, a normalized query would be preferred. Since we only care about the ORDER BY part, the code to normalize the query turned out like this abomination:

```ruby
PgQuery::Deparse.send(:deparse_select, { 'sortClause' => parsed_query_expect.tree[0]['RawStmt']['stmt']['SelectStmt']['sortClause'] })
```

A much cleaner way would be something like this (I'm skipping the `[0]` here for increased readability):

```ruby
parsed_query_expect.order_by.deparse
```

Another possibility here would be to add custom outputters to the object. This is an example of a possible syntax getting a block and registering this as an output module for all table names, effectively solving #85 without getting any MSSQL-specific code in a Pg-based project)

```ruby
parsed_query.table_name_output { |table| "[#{table}]" }
```

Another (maybe far fetched) possiblity would be to convert an SQL filter to an LDAP filter.

**Comment by lfittl** [comment]
@herwinw I think thats a very interesting idea, but as you say it would be a heavy restructuring.

I wonder if to start, you could build a sample case that converts the internal hash structure to a object-based structure for common SELECT queries?

That would allow us to validate whether this makes sense to support generally. Note that the hash structure itself comes from the JSON output from [libpg_query](http://github.com/lfittl/libpg_query), so whatever happens, we'd need to maintain a mapping between that hash structure and objects.

**Comment by lfittl** [comment]
@herwinw This has now been implemented in the new 2.0 release, which uses full Ruby classes in the parse tree result. Thanks for opening this a few years ago!

See the README for how it works: https://github.com/pganalyze/pg_query#parsing-a-query
